File: v2.1.0/src-tauri/src/glob.rs
pub fn parse_patterns(input: &str) -> Vec<String> {
    input
        .split(|c: char| c == ',' || c == '\n')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect()
}

fn glob_to_regex(glob: &str) -> String {
    let mut re = String::from("^");
    let mut chars = glob.chars().peekable();
    while let Some(c) = chars.next() {
        match c {
            '*' => {
                if chars.peek() == Some(&'*') {
                    chars.next();
                    re.push_str(".*");
                } else {
                    re.push_str("[^/]*");
                }
            }
            '?' => re.push_str("[^/]"),
            '.' | '+' | '^' | '$' | '{' | '}' | '(' | ')' | '|' | '[' | ']' | '\\' => {
                re.push('\\');
                re.push(c);
            }
            other => re.push(other),
        }
    }
    re.push('$');
    re
}

pub fn matches(rel_path: &str, patterns: &[String]) -> bool {
    if patterns.is_empty() {
        return false;
    }
    let normalized = rel_path.replace('\\', "/");
    let basename = normalized.rsplit('/').next().unwrap_or("");
    let mut excluded = false;
    for pattern in patterns {
        let (body, negated) = if let Some(body) = pattern.strip_prefix('!') {
            (body, true)
        } else {
            (pattern.as_str(), false)
        };
        let re = regex::Regex::new(&glob_to_regex(body));
        let Ok(re) = re else { continue };
        if re.is_match(&normalized) || re.is_match(basename) {
            excluded = !negated;
        }
    }
    excluded
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_separators() {
        assert_eq!(
            parse_patterns("*.tmp, node_modules\n.git"),
            vec!["*.tmp", "node_modules", ".git"]
        );
    }

    #[test]
    fn basename_match() {
        assert!(matches("foo/bar.tmp", &vec!["*.tmp".into()]));
        assert!(!matches("foo/bar.txt", &vec!["*.tmp".into()]));
    }

    #[test]
    fn negation_reincludes() {
        let p = vec!["*.tmp".into(), "!keep.tmp".into()];
        assert!(matches("skip.tmp", &p));
        assert!(!matches("keep.tmp", &p));
    }
}

File: v0.5.0-beta/src/lib/glob.js
function globToRegex(glob) {
  let re = '';
  for (let i = 0; i < glob.length; i++) {
    const c = glob[i];
    if (c === '*') {
      if (glob[i + 1] === '*') {
        re += '.*';
        i++;
      } else re += '[^/]*';
    } else if (c === '?') re += '[^/]';
    else if ('.+^${}()|[]\\'.includes(c)) re += '\\' + c;
    else re += c;
  }
  return new RegExp(`^${re}$`);
}

export function parsePatterns(input) {
  if (!input || typeof input !== 'string') return [];
  return input
    .split(/[,\n]/)
    .map((p) => p.trim())
    .filter(Boolean);
}

export function matches(relPath, patterns) {
  if (!patterns || patterns.length === 0) return false;
  const normalized = String(relPath).replace(/\\/g, '/');
  const basename = normalized.split('/').pop();
  let excluded = false;
  for (const pattern of patterns) {
    const negated = pattern.startsWith('!');
    const body = negated ? pattern.slice(1) : pattern;
    const regex = globToRegex(body);
    const hit = regex.test(normalized) || regex.test(basename);
    if (!hit) continue;
    excluded = !negated;
  }
  return excluded;
}
